lib/api/apiUtils/object/versioning.js

```diff
@@ -142,44 +188,98 @@ export function versioningPreprocessing(bucketName, bucketMD, objectKey,
             if (nullVersionId === undefined) {
                 return callback(null, options);
             }
-            options.deleteNullVersionData = true;
             return _deleteNullVersionMD(bucketName, objectKey,
-                { versionId: nullVersionId }, log,
-                err => callback(err, options));
+                { versionId: nullVersionId }, log, (err, nullDataToDelete) => {
+                    if (err) {
+                        process.stdout.write('====================\n');
+                        process.stdout.write('unexpected error\n');
+                        process.stdout.write(`err: ${err}\n`);
+                        process.stdout.write('method: versioningPreprocessing' +
+                        '0\n');
+                        log.warn('unexpected error deleting null version md', {
+                            error: err,
+                            method: 'versioningPreprocessing',
+                        });
+                        if (err === errors.NoSuchKey) {
+                            // it's possible there was a concurrent request to
+                            // delete the null version, so proceed with putting
+                            // a new version
+                            return callback(null, options);
+                        }
+                        return callback(errors.InternalError);
+                    }
+                    options.dataToDelete = nullDataToDelete;
+                    return callback(err, options);
+                });
         }
         // versioning is enabled, put the new version
         options.versioning = true;
         options.nullVersionId = nullVersionId;
         return callback(null, options);
+    // NOTE: Currently we do not accept requests with PUTs to specific
+    // version ID, but this may change after implementing georeplication.
     } else if (!mstVersionId) {
         // version-specific versioning operation, master is not versioned
         if (vstat === 'Suspended' || reqVersionId === 'null') {
             // object does not exist or is not versioned (before versioning)
             options.versionId = '';
             options.isNull = true;
-            options.deleteData = true;
+            options.dataToDelete = mstObjLocation;
             return callback(null, options);
         }
-        // TODO check AWS behaviour
         return callback(errors.BadRequest);
     } else if (mstIsNull) {
         // master is versioned and is a null version
         if (reqVersionId === 'null') {
             // overwrite the existing version, make new version null
             options.versionId = '';
             options.isNull = true;
-            options.deleteData = true;
+            options.dataToDelete = mstObjLocation;
             return callback(null, options);
         }
```

Comment:
I removed the duplicate code. I agree with your comment about the length, but at the moment I feel there are more important refactors to prioritize. :)